datafusion/optimizer/src/analyzer/type_coercion.rs

```diff
@@ -151,9 +151,27 @@ impl<'a> TypeCoercionRewriter<'a> {
             })
             .collect::<Result<Vec<_>>>()?;
 
+        // Join filter must be boolean
+        join.filter = join
+            .filter
+            .map(|expr| self.coerce_join_filter(expr))
+            .transpose()?;
+
         Ok(LogicalPlan::Join(join))
     }
 
+    fn coerce_join_filter(&self, expr: Expr) -> Result<Expr> {
+        let expr_type = expr.get_type(self.schema)?;
+        match expr_type {
+            DataType::Boolean => Ok(expr),
+            DataType::Null => Ok(Expr::Cast(expr::Cast::new(
+                Box::new(expr),
+                DataType::Boolean,
+            ))),
```

Comment:
I think this could be more concise if we used `Expr::cast_to, like this: 

```suggestion
            DataType::Null => expr.cast_to(&DataType::Boolean, self.schema),
```

Reply:
I agree, thank you @alamb